### PR TITLE
Gazebo vision plugin ROS

### PR DESCRIPTION
**Pull request by EliaTarasov**
This wip would add the possibility to subscribe to ROS external vision pose topic.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
The intent looks good, but obviously, the IFDEFs need work. Looking forward to it.

**Comment by EliaTarasov** [comment]
@TSC21  Previously in #215 you noted

> I think that the idea would be to actually add this capability on the gazebo_vision_plugin side by adding the possibility of subscribing to a ROS topic - though building this part should be dependent on a build env variable, as we have BUILD_ROS_INTERFACE on the CMakeLists.txt

Im not very familiar with adding env variable to PX4 build script, could you give any example on it?
Thanks!

**Comment by EliaTarasov** [comment]
@LorenzMeier  

> The intent looks good, but obviously, the IFDEFs need work

Yes it's just hint that needs to be replaced by some env variable.

File: CMakeLists.txt
Comment (TSC21) [inline CMakeLists.txt:278]: If you take this out of here, the vision plugin won't be able to build without the `BUILD_ROS_INTERFACE` set. The idea in this case is that you add `if (NOT ${BUILD_ROS_INTERFACE})`.

File: include/gazebo_vision_plugin.h
Comment (TSC21) [inline include/gazebo_vision_plugin.h]: Use [`add_definitions()`](https://cmake.org/cmake/help/v3.0/command/add_definitions.html) on the `CMakeLists.txt`:
```make
add_definitions(-DBUILD_ROS_INTERFACE=${BUILD_ROS_INTERFACE})
```
Then you can use them on code:
```c++
#if BUILD_ROS_INTERFACE == ON // or 1, not sure
// ROS Topic subscriber
#include <thread>
#include <ros/ros.h>
...
#endif
```

File: CMakeLists.txt
Comment (TSC21) [reply CMakeLists.txt:278]: Oh, sorry. Just saw you have it here https://github.com/PX4/sitl_gazebo/pull/223/files#diff-af3b638bc2a3e6c650974192a53c7291R344.

File: include/gazebo_vision_plugin.h
Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: Thanks, it works.

**Comment by TSC21** [comment]
Something's wrong with the CI on MacOS. Somehow is not able to detect Gazebo installed. Have to look further.

**Review by TSC21 — COMMENTED** [review]
I see there's a receiver for the data. How are you sending the data to the external VO system? Are you using the camera plugin? Can you add that to the iris_vision SDF?

**Comment by EliaTarasov** [comment]
Yes i send data using camera plugin. I will add this cam to iris_vision a little bit later, currently im working on my existing model.

File: include/gazebo_vision_plugin.h
Comment (TSC21) [inline include/gazebo_vision_plugin.h]: This should be a common topic name. Example: `"/visual/odometry"`. Then one should do a topic remap if needed. Also, options should be added for the topic msg type. Currently you just give the option for `geometry_msgs::PoseWithCovarianceStamped` but it should be open for `geometry_msgs::PoseStamped` and `nav_msgs::Odometry`.

Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: > This should be a common topic name. Example: "/visual/odometry". Then one should do a topic remap if needed.

Agree.

> Also, options should be added for the topic msg type. Currently you just give the option for geometry_msgs::PoseWithCovarianceStamped but it should be open for geometry_msgs::PoseStamped and nav_msgs::Odometry

Do you mean overloaded callback for such topics message types?
The problem here is to how to test these cases if only one type of message topic is posted by external VO.
For the geometry_msgs::PoseStamped it is possible to write the node that converts to it from geometry_msgs::PoseWithCovarianceStamped.
But i don't have a sensor that could completely fills nav_msgs::Odometry.

**Comment by TSC21** [comment]
@EliaTarasov something isn't right on your approach - CI is reporting that, with Cmake, it tries to build the ROS related parts - https://travis-ci.org/PX4/sitl_gazebo/jobs/395089239#L646. Can you please verify what's causing this?

**Comment by EliaTarasov** [comment]
Fixed. `BUILD_ROS_INTERFACE` eventually must be equal to `1` or `0` (and not to `ON` or `OFF`)

File: include/gazebo_vision_plugin.h
Comment (TSC21) [reply include/gazebo_vision_plugin.h]: What I mean is that you should have a callback for those msgs, even if you don't have a framework that does send those msgs. There are a lot out there that are able to do it, so this becomes more flexible. Check for example what's done in MAVROS. Basically you add a callback for each topic, and then the data is handled on the same method, basically filling the fields of the `odometry` Gazebo msg that it is able to so. If the callback is for `navigation_msgs/Odometry.msg` type, then you will be able to fill more that. If it is just `Pose` data, then you will fill less data. Will be better on that case to fill with NaNs, which in the case of the ODOMETRY Mavlink msg, for now, would mean 0's (Have to update the msg spec so we are able to state which data is valid or not).

**Comment by TSC21** [comment]
@EliaTarasov please add `vision_cam.sdf` as a workaround to the schema validation failing https://github.com/PX4/sitl_gazebo/blob/master/scripts/validate_sdf.bash#L57.
This is starting to look good. When are you able to show us something working on a screen record or so?
Also, can you please address https://github.com/PX4/sitl_gazebo/pull/223#discussion_r197545693? Thanks.

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: As you can see, most of the parsing code is the same between callbacks. A function to parse the same code would be a more wise approach.

**Review by TSC21 — CHANGES_REQUESTED** [review]

File: include/gazebo_vision_plugin.h
Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: I've added handlers, but not subscribers. The reason is that if multiple handlers with different arguments are subscribed to the one topic then we will get ROS error. I leave this problem for now.

Comment (TSC21) [reply include/gazebo_vision_plugin.h]: So why not different subscribers? You do need different handler per subscriber.

Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: Im not quite sure. Suppose we have Odometry handler and PoseStamped handler and they both are subscribed to "/visual/odometry" topic. If VO framework advertises topic with Odometry type, then PoseStamped handler crashes because it has different type. But i may be wrong i will check it.

Comment (TSC21) [reply include/gazebo_vision_plugin.h]:
```c++
if(sdf->HasElement("rosPoseSubTopic")) {
    _ros_pose_sub_topic = sdf->GetElement("rosPoseSubTopic")->Get<std::string>();
  } else {
    _ros_pose_sub_topic = kDefaultRosPoseTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS PoseStamped subscription to topic " << _ros_pose_sub_topic << "\n";
  }

if(sdf->HasElement("rosPoseCovSubTopic")) {
    _ros_posecov_sub_topic = sdf->GetElement("rosPoseCovSubTopic")->Get<std::string>();
  } else {
    _ros_posecov_sub_topic = kDefaultRosPoseCovTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS PoseWithCovarianceStamped subscription to topic " << _ros_posecov_sub_topic << "\n";
  }

if(sdf->HasElement("rosOdomSubTopic")) {
    _ros_odom_sub_topic = sdf->GetElement("rosOdomSubTopic")->Get<std::string>();
  } else {
    _ros_posecov_sub_topic = kDefaultRosOdomTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS Odometry subscription to topic " << _ros_odom_sub_topic << "\n";
  }

(...)

ros::SubscribeOptions so_pose = ros::SubscribeOptions::create<geometry_msgs::PoseStamped>(_ros_pose_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
ros::SubscribeOptions so_posecov = ros::SubscribeOptions::create<geometry_msgs::PoseWithCovarianceStamped>(_ros_posecov_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCovCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
ros::SubscribeOptions so_odom = ros::SubscribeOptions::create<navigation_msgs::Odometry>(_ros_odom_sub_topic, 1, boost::bind(&VisionPlugin::rosOdomCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);

this->_ros_pose_sub = this->_ros_node->subscribe(so_pose);
this->_ros_posecov_sub = this->_ros_node->subscribe(so_posecov);
this->_ros_odom_sub = this->_ros_node->subscribe(so_odom);

(...)
```

File: include/gazebo_vision_plugin.h
Comment (TSC21) [inline include/gazebo_vision_plugin.h]: why is this needed?

File: include/gazebo_vision_plugin.h
Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: I see your point. In this case problem disappears of course.

File: scripts/validate_sdf.bash
Comment (TSC21) [inline scripts/validate_sdf.bash]: Please put it in the same line as `fpv_cam.sdf`

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp:92]: I don't see this being used anywhere. Do you still want it here?

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: You can put all the bellow code on a specific function:
```c++
void VisionPlugin::handle_vision_data(geometry_msgs::Vector3 &position, geometry_msgs::Quaternion &orientation, geometry_msgs::Vector3 &lin_vel, geometry_msgs::Vector3 &ang_vel, float64[36] &pose_covariance, float64[36] &twist_covariance) {
    // handle all the parsing to gazebo msgs here
    (...)
}

void VisionPlugin::rosPoseCovCallBack(const geometry_msgs::PoseWithCovarianceStampedConstPtr& poseCovMsg)
{
    geometry_msgs::Vector3 lin_vel(0,0,0);
    geometry_msgs::Vector3 ang_vel(0,0,0);
    float64[36] twist_covariance = { }; // all values are zero
    handle_vision_data(poseCov->pose.pose.position, poseCov->pose.pose.orientation, lin_vel, ang_vel, poseCov->pose.covariance, twist_covariance);
}

// then the same for the other Callbacks
```

**Review by TSC21 — CHANGES_REQUESTED** [review]

File: include/gazebo_vision_plugin.h
Comment (EliaTarasov) [reply include/gazebo_vision_plugin.h]: I will remove only the last string. Others are needed for callbacks to be compiled.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp:92]: Please look at https://github.com/EliaTarasov/sitl_gazebo/blob/7c0526d73d4d4dcba9d1baaa4c7b4c920b496105/src/gazebo_vision_plugin.cpp#L132
It is used in case when plugins are built with ROS but this particular plugin is need to be used without ROS.

File: scripts/validate_sdf.bash
Comment (EliaTarasov) [reply scripts/validate_sdf.bash]: Done

**Comment by EliaTarasov** [comment]
> When are you able to show us something working on a screen record or so?


I've added screen record to show how the simple VO flight does look [like](https://drive.google.com/open?id=1YDarUp7UO9mSssxAfz61z6jpslsYgt-r)

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Agree.

**Comment by TSC21** [comment]
> I've added screen record to show how the simple VO flight does look [like](https://drive.google.com/open?id=1YDarUp7UO9mSssxAfz61z6jpslsYgt-r)

Pretty cool stuff!

**Comment by TSC21** [comment]
@EliaTarasov when this PR is finished would you mind add some documentation about the existence of, purpose, fundamentals and usage of the ROS interface with external vision pose estimation sources in a new section or integrated in one in https://dev.px4.io/en/ros/? @hamishwillee what's your opinion about it?

**Comment by hamishwillee** [comment]
> @EliaTarasov when this PR is finished would you mind add some documentation about the existence of, purpose, fundamentals and usage of the ROS interface with external vision pose estimation sources in a new section or integrated in one in https://dev.px4.io/en/ros/? @hamishwillee what's your opinion about it?

Without some information about the purpose and where it "sits" it is hard to know exactly where this should be integrated. My first thought is this should be part of a (very much updated) https://dev.px4.io/en/ros/external_position_estimation.html

**Comment by TSC21** [comment]
> My first thought is this should be part of a (very much updated) https://dev.px4.io/en/ros/external_position_estimation.html

Yeah I think that fits!

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: Please put all on the same line

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: Let's use Eigen instead. More flexible than ROS types:
```c++
void VisionPlugin::handle_vision_data(uint64_t &usec, Eigen::Vector3d pos, Eigen::Quaterniond &q, Eigen::Vector3d lin_vel, Eigen::Vector3d ang_vel, std::array<double, 36> &pose_cov, std::array<double, 36> &twist_cov) {
    ....
}
```

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: if you use `std::array`, then you can use `std::copy`.

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]:
```c++
std::array<double, 36> pose_covariance{};
std::array<double, 36> twist_covariance{};
```

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]:
```c++
std::copy(odomMsg->pose.covariance.data().begin(), odomMsg->pose.covariance.data().end(), pose_covariance.begin());
```

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: Same as the previous comments

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: `odomMsg->header.stamp.toNSec(),`

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: same

**Review by TSC21 — CHANGES_REQUESTED** [review]
Almost there @EliaTarasov. Just needs some more adjustments. Thanks for your contributions so far!

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Doesn't work with Gazebo Odometry. If i change it to, for instance:
`std::copy(pose_cov.begin(), pose_cov.end(), odom_msg.pose_covariance.begin());
    std::copy(twist_cov.begin(), twist_cov.end(), odom_msg.twist_covariance.begin()); `
then it gives the error:
`‘((gazebo::VisionPlugin*)this)->gazebo::VisionPlugin::odom_msg.nav_msgs::msgs::Odometry::pose_covariance’ does not have class type
     std::copy(pose_cov.begin(), pose_cov.end(), odom_msg.pose_covariance.begin()); `

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

File: src/gazebo_vision_plugin.cpp
Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Not sure. https://github.com/PX4/sitl_gazebo/blob/master/msgs/Odometry.proto#L8 shows that timestamp is in microsecs, while `odomMsg->header.stamp.toNSec()` gives nanosecs.

Comment (TSC21) [reply src/gazebo_vision_plugin.cpp]: @EliaTarasov you didn't change `toSec()` to `toNSec()`. If use it, you don't have to do the conversion on the method above.

Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: @TSC21 can you please clarify my comment above?

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp:193]: Actually, in terms of usability, this should be propagated in nanosec and then converted to microsec on the mavlink interface. Would you mind changing the proto of the Odometry msg? Basically changing `usec` field to `timestamp` and have it progated in nanosec. Thanks!

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [inline src/gazebo_vision_plugin.cpp]: Check `eigen_conversions`: http://docs.ros.org/groovy/api/eigen_conversions/html/namespacetf.html.

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [reply src/gazebo_vision_plugin.cpp]: Yes. The idea is to change this and propate the timestamp in nanosecs. Check https://github.com/PX4/sitl_gazebo/pull/223#discussion_r198786207. So basically the proto needs to be updated (didn't thought of it back when it was created).

File: src/gazebo_vision_plugin.cpp
Comment (TSC21) [reply src/gazebo_vision_plugin.cpp]: `tf::vectorMsgToEigen()` and `tf::quaternionMsgToEigen()`.

Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: > tf::vectorMsgToEigen() and tf::quaternionMsgToEigen()

Yes i found this but now im looking for header to include this stuff

Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: It's ok: `#include <eigen_conversions/eigen_msg.h>`

Comment (EliaTarasov) [reply src/gazebo_vision_plugin.cpp]: Done.

**Review by TSC21 — APPROVED** [review]
This looks good. I will add the nsec timestamp after. @EliaTarasov do you mind adding some doc about this on the Devguide?

**Comment by EliaTarasov** [comment]
@TSC21 
>  I will add the nsec timestamp after.

Thanks a lot, this would be much needed.

> This looks good

I double checked all stuff and found that CMake [issue](https://github.com/EliaTarasov/sitl_gazebo/blob/372513df4379a44e89aaadb1ac0100f65f156ce6/include/gazebo_vision_plugin.h#L59) with `BUILD_ROS_INTERFACE` remains.
It doesn't matter either [BUILD_ROS_INTERFACE](https://github.com/EliaTarasov/sitl_gazebo/blob/c7f0dcefe5903bd456094fd7cbb5b888023130ea/CMakeLists.txt#L25) is set to `ON` or is set to `OFF` inside CMake.
1. If `#if BUILD_ROS_INTERFACE == 1 ... #endif` is used then macro depended code is never build.
2. If `#if BUILD_ROS_INTERFACE == ON ... #endif` is used then macro depended code is always build.

Adding  `add_definitions(-DBUILD_ROS_INTERFACE=${BUILD_ROS_INTERFACE})` does not make any difference.

The code itself works ok.

This issue must be addressed before merge i guess.
Currently i'm looking for solution.

**Comment by mhkabir** [comment]
Question : what is the intent behind this plugin? What does it add, compared to publishing Gazebo images to ROS and feeding back odometry estimates through mavros?

**Comment by TSC21** [comment]
> Question : what is the intent behind this plugin? What does it add, compared to publishing Gazebo images to ROS and feeding back odometry estimates through mavros?

Simple: not depending on MAVROS.

**Comment by TSC21** [comment]
> This issue must be addressed before merge i guess.
Currently i'm looking for solution.

OK @EliaTarasov thanks for the feedback on that. Find it awkward not functioning as it should. have you tried `target_compile_definitions()` instead?

**Comment by TSC21** [comment]
So using `#if BUILD_ROS_INTERFACE == ON ... #endif`, if you do `catkin build -DBUILD_ROS_INTERFACE=OFF` or just `catkin build`, does it build or not?

**Review by mhkabir — CHANGES_REQUESTED** [review]
What's the point of not depending on mavros if you're depending on ROS anyway for everything else like image I/O. This interface will just be adding a maintenance burden, and making changes/improvements in either mavros or here will have to be duplicated in the other one. 

I'm not convinced unless there is a more compelling reason or use-case.

**Comment by EliaTarasov** [comment]
@TSC21 

>  Find it awkward not functioning as it should. have you tried target_compile_definitions() instead?

I will try it ASAP.

> So using #if BUILD_ROS_INTERFACE == ON ... #endif, if you do catkin build -DBUILD_ROS_INTERFACE=OFF or just catkin build, does it build or not?

It does build and work as in the [example](https://drive.google.com/file/d/1YDarUp7UO9mSssxAfz61z6jpslsYgt-r/view)

**Comment by TSC21** [comment]
@mhkabir this and MAVROS are different projects. The approach taken is maintanable, with a small footprint and not a maintenance burden as you say. My evaluation is that this is a good approach and adds few dependencies - adding a dependency on MAVROS would mean that we would depend on MAVROS API, and a change in one side would mean a change on the other, while like this it wouldn't.

**Review by TSC21 — COMMENTED** [review]
@EliaTarasov does this solve the issue?

**Comment by EliaTarasov** [comment]
@TSC21 Yes it does. I checked it several times and now it is fixed.
Thanks for the help!

**Review by TSC21 — APPROVED** [review]

**Comment by TSC21** [comment]
Thanks Elia. Would you mind adding some description about what this does and how to use it on the PX4 devguide?

**Comment by EliaTarasov** [comment]
Yes but i need some time to figure out how to actually add documentation and also to formulate key ideas and express it.

**Comment by EliaTarasov** [comment]
@mhkabir The advantage is also to provide more consisted use of the `vision_plugin`. Currently it is necessary to comment out or delete  `vision_plugin` from the `sdf` model if external VO feeds through mavros.

It is convenient to debug VO camera's orientation without bringing mavros rotations into view, considering the fact that gazebo itself has many options for childs links rotations. In case of more or less complex model the process of finding proper orientation could be tricky even without mavros.

If this solution works out of the box it is less time consuming to create own vision model.

**Comment by TSC21** [comment]
> Yes but i need some time to figure out how to actually add documentation and also to formulate key ideas and express it.

You would start from [here](https://github.com/PX4/Devguide). Take your time to write it. Given the importance of this for VIO sim testing, it will be important to have it documented.

**Comment by EliaTarasov** [comment]
Ok, got it.

**Comment by TSC21** [comment]
Hey @EliaTarasov! Any updates on the documentation side?

**Comment by EliaTarasov** [comment]
@TSC21 Still not. By the way - how do you think, is it worth to prepare documentation given the fact that documented code probably will not be merged?

**Comment by TSC21** [comment]
@EliaTarasov which code are you talking about? The one from this PR?

**Comment by EliaTarasov** [comment]
> @EliaTarasov which code are you talking about? The one from this PR?

Yes.

**Comment by TSC21** [comment]
Let's wait for @LorenzMeier to unblock this.

**Comment by TSC21** [comment]
@LorenzMeier can you please unblock this? This will be important if we want to test this with VIO estimation sources, like ROVIO, in simulated environments - which can be plugged on our CI.

**Comment by LorenzMeier** [comment]
@EliaTarasov It took me some time to wrap my head around this, but I think I have now. What is the reason for not just using MAVROS in a launch file? After all, you need to configure the VIO pipeline already. And you need MAVROS still on the vehicle. So what do you win by using it here without MAVROS?

**Comment by EliaTarasov** [comment]
@LorenzMeier The inital reason was to add option that allows to use `vision_plugin` in a more flexible way. Currently `vision_plugin` doesn't simulate VO and to use external VO you need to remove `vision_plugin` from `sdf` model. It would be better to keep `vision_plugin` in every `vision` SITL model but add this option to change it's parameters to use external VO.

**Comment by EliaTarasov** [comment]
@TSC21 What do you think?

**Comment by TSC21** [comment]
@EliaTarasov I asked @LorenzMeier to untie this, so you know my opinion. Though I do understand that we already have MAVROS to make the bridging, we will always be dependent on MAVROS API so this interface can happen, and that's what I wanted to avoid. Since the opinions here diverge, I guess it will be @LorenzMeier decision to get this further or not.
In any case, this interface presents itself as an alternative now for what is coming next, which will be having RTPS interfacing with the autopilot in order to exchange VIO data. So being it through Gazebo or through MAVROS, they will both be deprecated in the sense that the RTPS solution would be the most adequate one for VIO interfacing (among others).

**Pull request closed without merging**
